### PR TITLE
8269022: Put evacuation failure string directly into gc=info log message

### PR DESCRIPTION
**Pull request by tschatzl**
Hi all,

  can I have reviews for this change that puts the evacuation failure marker from a separate log line into the GC timing message. I.e. instead of

```
Pause Young (Normal) (G1 Evacuation Pause)
To-space exhausted
```
into

```
Pause Young (Normal) (Evacuation Failure) (G1 Evacuation Pause) 
```

This is imho easier to read, better to process and less wasteful with log space.

Testing: tier1, gc/g1 tests

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8269022](https://bugs.openjdk.java.net/browse/JDK-8269022): Put evacuation failure string directly into gc=info log message

### Reviewers
 * [Ivan Walulya](https://openjdk.java.net/census#iwalulya) (@walulyai - Committer) ⚠️ Review applies to 298eb0a99d2bdef0f7e586f2bb7b460a59f28df3
 * [Leo Korinth](https://openjdk.java.net/census#lkorinth) (@lkorinth - Committer) ⚠️ Review applies to 004c0a9c82a52637465c6d648862fc73bd2daacc
 * [Kim Barrett](https://openjdk.java.net/census#kbarrett) (@kimbarrett - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4539/head:pull/4539` \
`$ git checkout pull/4539`

Update a local copy of the PR: \
`$ git checkout pull/4539` \
`$ git pull https://git.openjdk.java.net/jdk pull/4539/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4539`

View PR using the GUI difftool: \
`$ git pr show -t 4539`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4539.diff">https://git.openjdk.java.net/jdk/pull/4539.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/gc/g1/g1CollectedHeap.cpp [hunk 1]
```diff
@@ -2915,6 +2904,40 @@ void G1CollectedHeap::gc_tracer_report_gc_end(bool concurrent_operation_is_full_
...
+    // Take snapshot of current pause type at start as it may be modified during gc.
+    // The strings for all Concurrent Start pauses are the same, so the parameter
+    // does not matter here.
+    _pause_type(G1CollectedHeap::heap()->collector_state()->young_gc_pause_type(false /* concurrent_operation_is_full_mark */)),
+    _tt(update_young_gc_name(), NULL, cause, true) {
+  }
+
+  ~G1YoungGCTraceTime() {
+    update_young_gc_name();
+  }
```
Comment (lkorinth) [inline src/hotspot/share/gc/g1/g1CollectedHeap.cpp:2938]: Do I understand correctly that we update the name in _young_gc_name_data in the destructor, and then depend on that _tt does not  copy that value earlier --- only the pointer? This seems quite fragile if GCTraceTimeWrapper would change, and I think would at least warrant a comment in the constructor that the string *will* be updated in the destructor.

File: test/hotspot/jtreg/gc/g1/TestEvacuationFailure.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        output.shouldHaveExitValue(0);
+    }
+
+    static class GCTestWithEvacuationFailure {
+        private static byte[] garbage;
+        private static byte[] largeObject;
+        private static Object[] holder = new Object[200]; // Must be larger than G1EvacuationFailureALotCount
+
+        public static void main(String [] args) {
+            largeObject = new byte[16*1024*1024];
```
Comment (lkorinth) [inline test/hotspot/jtreg/gc/g1/TestEvacuationFailure.java:70]: space around asterisks

**Review by walulyai — APPROVED** [review]

File: src/hotspot/share/gc/g1/g1CollectedHeap.cpp [hunk 3]
```diff
@@ -2915,6 +2904,40 @@ void G1CollectedHeap::gc_tracer_report_gc_end(bool concurrent_operation_is_full_
...
+// GC cause.
+class G1YoungGCTraceTime {
+  G1GCPauseType _pause_type;
+
+  static const uint MaxYoungGCNameLength = 128;
+  char _young_gc_name_data[MaxYoungGCNameLength];
+
+  GCTraceTime(Info, gc) _tt;
+
+  char* update_young_gc_name() {
```
Comment (lkorinth) [inline src/hotspot/share/gc/g1/g1CollectedHeap.cpp:2917]: this can be made a const char*, right?

**Review by lkorinth — APPROVED** [review]
I have some comments, but otherwise it looks good to me.

File: src/hotspot/share/gc/g1/g1CollectedHeap.cpp (hunk 1 above)
Comment (tschatzl) [reply src/hotspot/share/gc/g1/g1CollectedHeap.cpp:2938]: Added. Note that the previous code also had that assumption, this is why the `young_gc_name` array needed to be alive the whole method.
The added test will detect if it does not work any more.

**Review by lkorinth — APPROVED** [review]
Still looks good to me!

**Review by albertnetymk — COMMENTED** [review]
Regarding `Pause Young (Normal) (Evacuation Failure) (G1 Evacuation Pause)`, I interpret the three pairs of parens as:

1. (<gc pause kind, deciding what should be done in this pause>)
2. (<sth unexpected happened, deviating from the original plan>)
3. (<what initiates this pause>)

2 occurs strictly later than 1&3, so a more natural order is

`Pause Young (Normal) (G1 Evacuation Pause) (Evacuation Failure)`.

This way, the original structure (meaning of the first two pairs of parens) could be preserved in all cases. When sth unexpected happens, the log is only appended, leaving the "conceptually existing" part intact.

Ofc, this is very subjective; just my 2 cents. (Merely a comment, not a review. I am fine with the current change.)

**Comment by tschatzl** [comment]
I am aware of this and this has come up before. The problem is that the order of adding these strings is fixed by the execution order of the scoped objects. My initial attempt to fix this was too messy, but 'll look into this again, maybe I can find some better way.

**Comment by tschatzl** [comment]
@albertnetymk : fixed :)

**Review by kimbarrett — APPROVED** [review]
Looks good.

I agreed with Albert's comment about the order of the annotations.  I'm glad you found a (nice, not kludgy) way to fix that.

**Comment by tschatzl** [comment]
Thanks @albertnetymk @walulyai @lkorinth @kimbarrett for your reviews.

/integrate

**Pull request closed without merging**